### PR TITLE
ISSUE #40: LightGBM Training, Model Conversion & Testing

### PR DESCRIPTION
**Pull request by venkate5hgunda**
[1] Training
[1] Conversion for ONNX
[1] Conversion for HummingBird (No changed needed)
[1] Conversion for TreeLite (Added Support)
[1] Conversion for LLeaves (Coding Done. Test Pending)
[0] Test for Sk-Learn
[0] Test for ONNX
[0] Test for HummingBird Backends
[0] Test for TreeLite
[0] Test for LLeaves
[0] Experimental: See if TFDF GradientBoostedTrees Class supports LightGBM

### CODE REVIEW

File: model-inference/decisionTree/experiments/convert_trained_model_to_framework.py [hunk 1]
```diff
@@ -21,27 +21,28 @@
...
 def parse_arguments():
+    # check_argument_conflicts(args)  # TODO: Move this function from the bottom to here, after checking with Prof.
     global DATASET, MODEL, FRAMEWORKS
     parser = argparse.ArgumentParser(description='Arguments for train_model.')
     parser.add_argument("-d", "--dataset", type=str, choices=['higgs', 'airline_classification', 'airline_regression', 'fraud', 'year', 'epsilon'],
         help="Dataset to be trained. Choose from ['higgs', 'airline_classification', 'airline_regression', 'fraud', 'year', 'epsilon']")
-    parser.add_argument("-m", "--model", type=str, choices=['randomforest', 'xgboost'],
-        help="Model name. Choose from ['randomforest', 'xgboost']")
+    parser.add_argument("-m", "--model", type=str, choices=['randomforest', 'xgboost', 'lightgbm'],
+        help="Model name. Choose from ['randomforest', 'xgboost', 'lighgbm]")
```
Comment (venkate5hgunda) [inline model-inference/decisionTree/experiments/convert_trained_model_to_framework.py:30]: - [ ] Correct this spelling mistake.

File: model-inference/decisionTree/experiments/convert_trained_model_to_framework.py [hunk 2]
```diff
@@ -126,11 +139,11 @@ def convert_to_onnx_model(model, config):
 def convert_to_treelite_model(model, config):
     #converting to TreeLite model
     #Prerequisite: install treelite (https://treelite.readthedocs.io/en/latest/install.html)
-    if MODEL == "xgboost":
+    if MODEL in {"xgboost", "lightgbm"}:
         treelite_time_start = time.time()
         model_path = relative2abspath("models", f"{DATASET}_{MODEL}_{config['num_trees']}_{config['depth']}.model")
-        model.save_model(model_path)
-        treelite_model = treelite.Model.load(model_path, model_format='xgboost')
+        model.save_model(model_path)  # TODO: Why are we saving this directly to a .model format?
```
Comment (hguan6) [inline model-inference/decisionTree/experiments/convert_trained_model_to_framework.py:145]: To distinguish xgboost and lightgbm, we can name them with different file extension

File: model-inference/decisionTree/experiments/convert_trained_model_to_framework.py [hunk 3]
```diff
@@ -143,7 +156,24 @@ def convert_to_treelite_model(model, config):
         # print("predictions", pred_onx[0])
         # # print("predict_proba", pred_onx[1][:1])
     else:
-        print(f"TreeLite models only supports xgboost algorithm, but does not support {MODEL} algorithm.")
+        print(f"TreeLite only supports conversion from xgboost, lightgbm and sk-learn based models. Does not support {MODEL}.")
+
+
+def convert_to_lleaves_model(model, config):
```
Comment (hguan6) [inline model-inference/decisionTree/experiments/convert_trained_model_to_framework.py:162]: Where is this function called?

Comment (venkate5hgunda) [reply model-inference/decisionTree/experiments/convert_trained_model_to_framework.py:165]: Added the reference code for this function.

File: model-inference/decisionTree/experiments/convert_trained_model_to_framework.py (hunk 2 above)
Comment (venkate5hgunda) [reply model-inference/decisionTree/experiments/convert_trained_model_to_framework.py]: I saw that Treelite Models have a .model extension. I understand how this is working, and will remove this TODO.

**Comment by jiazou-bigdata** [comment]
@venkate5hgunda 

python train_model.py -d higgs -m lightgbm doesn't work at my side. 
Could you please check? There is some problem related to -f. However, train_model.py will always train model using scikit-learn and should not rely on frameworks. Anything did I miss?

The complete log is as follows:

(base) ubuntu@ip-172-31-16-75:~/netsdb/model-inference/decisionTree/experiments$ git branch
  42-support-large-scale-decision-forest-decision-in-netsdb
  batch_experiments
* lightgbm-experiments
  master

(base) ubuntu@ip-172-31-16-75:~/netsdb/model-inference/decisionTree/experiments$ python train_model.py -d higgs -m lightgbm
Traceback (most recent call last):
  File "/home/ubuntu/netsdb/model-inference/decisionTree/experiments/train_model.py", line 124, in <module>
    parse_arguments()
  File "/home/ubuntu/netsdb/model-inference/decisionTree/experiments/train_model.py", line 29, in parse_arguments
    check_argument_conflicts(args)
  File "/home/ubuntu/netsdb/model-inference/decisionTree/experiments/model_helper.py", line 163, in check_argument_conflicts
    if "lleaves" in frameworks and not model == "lightgbm":
UnboundLocalError: local variable 'frameworks' referenced before assignment

**Comment by venkate5hgunda** [comment]
Hello Professor, apologies for that error. It was due to an indentation. Fixed it and pushed. I was waiting for model conversion to finish before pushing the code changes along with them.

**Comment by jiazou-bigdata** [comment]
> Hello Professor, apologies for that error. It was due to an indentation. Fixed it and pushed. I was waiting for model conversion to finish before pushing the code changes along with them.

Thanks for the quick fix. Now it worked from my side.

**Comment by jiazou-bigdata** [comment]
@venkate5hgunda 

Another observation is that Lleaves runs with single thread at my side. Is it true that Lleaves doesn't support multi-threading at this point?

**Comment by jiazou-bigdata** [comment]
@venkate5hgunda 
The Lleaves compilation process is very slow, and can only execute in single-thread. I posted an issue [here](https://github.com/siboehm/lleaves/issues/26) and you should also join the discussion with the Lleaves developers. 
The prediction should be fine, but I never get a chance to see the prediction performance for 500 and 1600 trees, because of the long compilation process. 

It could be a limitation of LLVM JIT code generation compared to TreeLite's source code generation. Both are slow in compiling the model, but the latter only requires compilation once and the .so library can be loaded dynamically, while the former requires compilation for each application. This is a great finding for our paper. Thanks Venkatesh!

Besides this, the PR looks great! Please resolve the conflicts and merge!

**Review by jiazou-bigdata — APPROVED** [review]

**Comment by venkate5hgunda** [comment]
Thank you, Prof @jiazou-bigdata.

At the moment, we're making use of the cache-path suggested by the author of Lleaves during model conversion (https://github.com/asu-cactus/netsdb/blob/master/model-inference/decisionTree/experiments/convert_trained_model_to_framework.py#L176)

But, as this is for benchmarking, and that we'd want that compilation time as well, did not use it for testing. Do you want me to add this to testing code?

**Comment by jiazou-bigdata** [comment]
> Thank you, Prof @jiazou-bigdata.
> 
> At the moment, we're making use of the cache-path suggested by the author of Lleaves during model conversion to save the ELF file (https://github.com/asu-cactus/netsdb/blob/master/model-inference/decisionTree/experiments/convert_trained_model_to_framework.py#L176)
> 
> But, as this is for benchmarking, and that we'd want that compilation time as well, did not use it for testing. Do you want me to add this to testing code?

Hi Venktash, sorry I just realized there is an option for caching the model. Then it is very similar to TreeLite in that the model can be compiled only once. Yes, could you please reopen this PR or create a new PR to change the benchmark code in test_model.py to load the model from the cached path without compilation? Thank you!

**Pull request merged**